### PR TITLE
Adding caching for clusters() query

### PR DESCRIPTION
**Pull request by arepraneeth**
If we have more clusters in grange.State , clusters() query will become expensive.
Adding caching for clusters() query results

### CODE REVIEW

File: evaluator.go [hunk 1]
```diff
@@ -73,12 +78,24 @@ func (s *State) Clusters() map[string]Cluster {
 	return s.clusters
 }
 
+// enable clusterQueryResultCaching
```
Comment (xaviershay) [inline evaluator.go]: does this really need to be configurable? Why can't it just be always on?

File: evaluator.go [hunk 2]
```diff
@@ -153,6 +171,31 @@ func (state *State) Query(input string) (Result, error) {
 	return evalRangeWithContext(input, state, &context)
 }
 
+// Used for populating state.cachedclusterQueryResult
+// Go over all clusters in state.clusters, look if element is part of cluster definition,
+// if yes, add it name of cluster to cachedCQR[element]
+// as a precaution, we donot want to cache empty results
```
Comment (xaviershay) [inline evaluator.go]: grammar nits: "add it name", "donot".

File: evaluator.go [hunk 3]
```diff
@@ -645,10 +692,47 @@ func (c *evalContext) addResult(value string) {
 	c.currentResult.Add(value)
 }
 
+func (c *evalContext) addSliceToResult(slice []interface{}) {
+	if c.currentResult.Cardinality()+len(slice) >= MaxResults {
+		panic(tooManyResults{})
```
Comment (xaviershay) [inline evaluator.go]: pretty sure, panic is inappropriate here. We don't panic anywhere else max results is exceeded. (I think we just truncate?)

File: evaluator.go [hunk 4]
```diff
@@ -527,13 +570,17 @@ func (n nodeFunction) visit(state *State, context *evalContext) error {
...
-		for clusterName, _ := range state.clusters {
-			subContext = context.subCluster(clusterName)
-			clusterLookup(state, &subContext, "CLUSTER")
-
-			for value := range subContext.resultIter() {
-				if lookingFor.Contains(value) {
-					context.addResult(clusterName)
+		if state.cacheClusterQueryResults {
+			results := state.getResultsFromCachedCQRforSlice(lookingFor.ToSlice())
+			context.addSliceToResult(results.ToSlice())
```
Comment (xaviershay) [inline evaluator.go]: Using slices feels weird to me, mostly because we don't have to do it anywhere else. Why is it necessary here? (Rather than just using set operations used elsewhere?)

File: evaluator.go [hunk 5]
```diff
@@ -645,10 +692,47 @@ func (c *evalContext) addResult(value string) {
...
+			// remove names we found in cache
+			sliceMap.Remove(name)
+		}
+	}
+	// return if we found all queries from cachedCQR
+	if sliceMap.Cardinality() == 0 {
+		return context.currentResult
+	}
+	// populate cachedCQR for leftOver elements in slice
+	state.populateCachedCQRforSlice(sliceMap.ToSlice())
```
Comment (xaviershay) [inline evaluator.go]: seems like making `populateCachedCQRforSlice` could be made cheaply idempotent for entries in the cache. Then, you don't need to do this double handling, you just populate the cache for everything (which will be noop for stuff already in it) then do a single loop over it.

File: evaluator_test.go [hunk 6]
```diff
@@ -231,6 +233,81 @@ func TestClusters(t *testing.T) {
 	}))
 }
 
+func TestClustersBoth(t *testing.T) {
```
Comment (xaviershay) [inline evaluator_test.go]: should this be in `range-spec` instead?

File: evaluator_test.go [hunk 7]
```diff
@@ -231,6 +233,81 @@ func TestClusters(t *testing.T) {
...
+	testEval(t, NewResult("a", "b", "c"), "clusters(one,three)", multiCluster(map[string]Cluster{
+		"a": Cluster{"CLUSTER": []string{"two", "one"}},
+		"b": Cluster{"CLUSTER": []string{"$ALL"}, "ALL": []string{"one"}},
+		"c": Cluster{"CLUSTER": []string{"three"}},
+	}))
+}
+
+func TestClusterQueryCache(t *testing.T) {
```
Comment (xaviershay) [inline evaluator_test.go]: WDYT about running range spec twice, once with enabled, once with disabled? (Though per above comment I'm not convinced it should be configurable)

File: evaluator_test.go [hunk 8]
```diff
@@ -231,6 +233,81 @@ func TestClusters(t *testing.T) {
...
+	state.Query("clusters(one)")
+	expected_length := 1
+	actual_length_of_keys := len(state.cachedCQR.Keys())
+	if actual_length_of_keys != expected_length {
+		t.Errorf("Expected '%d' Key(s) in 'state.cachedCQR', but got %d",
+			expected_length, actual_length_of_keys)
+	}
+}
+
+func TestAddValueToCachedCQR(t *testing.T) {
```
Comment (xaviershay) [inline evaluator_test.go]: Since this isn't a behaviour change, it shouldn't require any extra unit tests. Maybe you want to beef up range-spec because you found some edge cases, but in general a) the existing specs should be good enough to prove correctness, and b) a performance benchmark is proof that you actually made things faster. (Just reading off query times from log would satisfy me - doesn't need to be rigorous. Sounds like you've done that part already.)

File: evaluator.go [hunk 9]
```diff
@@ -153,6 +171,31 @@ func (state *State) Query(input string) (Result, error) {
 	return evalRangeWithContext(input, state, &context)
 }
 
+// Used for populating state.cachedclusterQueryResult
+// Go over all clusters in state.clusters, look if element is part of cluster definition,
+// if yes, add it name of cluster to cachedCQR[element]
+// as a precaution, we donot want to cache empty results
+func (state *State) populateCachedCQRforSlice(slice []interface{}) {
```
Comment (xaviershay) [inline evaluator.go]: `PrimeCache` should ensure this cache is populated, I don't see code that does that.

ref: https://github.com/square/grange/blob/master/evaluator.go#L111

File: evaluator.go (hunk 1 above)
Comment (arepraneeth) [inline evaluator.go]: was concerned about increase in memory requirements for caching results. Caching will be an opt in feature at expense of little memory increase

File: evaluator.go (hunk 4 above)
Comment (arepraneeth) [inline evaluator.go]: agree, will change it.

File: evaluator.go (hunk 1 above)
Comment (xaviershay) [inline evaluator.go]: any way to quantify it?

My experience with RSpec has left me really reluctant to expand configuration API without really good reason, since it gets unruly really quick.

File: evaluator.go (hunk 3 above)
Comment (arepraneeth) [inline evaluator.go]: `recover()` section of func `evalRangeInplace` would see `tooManyResults` , and returns results. This is same behavior in `addResult()` function

File: evaluator.go (hunk 5 above)
Comment (arepraneeth) [inline evaluator.go]: Pre populating cache would increase start up time, hence decided to populate cache as we query.  Need to add stats on how much time it takes to build cache and then will modify it accordingly

File: evaluator.go (hunk 1 above)
Comment (arepraneeth) [inline evaluator.go]: considering a scenario of 5000 clusters and 6000 nodes, and node length=50 bytes, cluster name length = 50bytes,
MemoryRequiredForCaching = `no.Nodes*sizeOfnode (memoryForKeys) + NumberOfNodes*NumberOfClustersNodeisPartOf*SizeOfCluster (memoryForValue)`

noOfClustersNodeIsPartOf | MemoryRequiredInMb
--------------------------- | ----------------------
1|0.6
10|3.3
50|15.3
100|30.3
1000|300.3
5000|1500.3

But common scenarios, a node is part of 10-50 clusters, so increase in memory would be around ~3-15Mb + memoryOverheadofManaging concurrentMaps another 15Mb?

File: evaluator_test.go (hunk 8 above)
Comment (arepraneeth) [inline evaluator_test.go]: removed

File: evaluator_test.go (hunk 7 above)
Comment (arepraneeth) [inline evaluator_test.go]: added test to check length of keys in cache without enabling caching.

File: evaluator_test.go (hunk 6 above)
Comment (arepraneeth) [inline evaluator_test.go]: moved # https://github.com/square/range-spec/pull/25

File: evaluator.go (hunk 4 above)
Comment (arepraneeth) [inline evaluator.go]: updated

File: evaluator.go [hunk 10]
```diff
@@ -645,10 +692,46 @@ func (c *evalContext) addResult(value string) {
...
+			context.addSetToResult(subResult.(Result))
+			// remove names we found in cache
+			set.Remove(name)
+		}
+	}
+	// return if we found all queries from cachedCQR
+	if set.Cardinality() == 0 {
+		return context.currentResult
+	}
+	// populate cachedCQR for leftOver elements in slice
```
Comment (xaviershay) [inline evaluator.go]: nit: no longer a slice :)

File: evaluator.go (hunk 1 above)
Comment (xaviershay) [inline evaluator.go]: let's leave out for now (always enabled) and consider adding in if anyone requests it

File: evaluator.go (hunk 9 above)
Comment (xaviershay) [inline evaluator.go]: If not going to fix `PrimeCache` now, please add a caveat to documentation for it that it doesn't actually do what it says it's going to do

File: evaluator.go (hunk 3 above)
Comment (xaviershay) [inline evaluator.go]: oh I didn't remember that part for the code :)

File: evaluator.go (hunk 5 above)
Comment (xaviershay) [inline evaluator.go]: I'm saying something different: this function is _already_ going to populate the cache for every item in `sliceMap`. So the code would be simpler if it was something like:

```
foreach slice in sliceMap
  populateCache(slice)
  fetchFromCache(slice)
```

If the cache is already populated for a `slice`, then it can just early exit. This way you don't need to keep track of `leftovers`

**Comment by arepraneeth** [comment]
Updated Pullrequest to include 
1) parallelism in PrimeCache()
2) exposed some of metrics related to cache and state

File: evaluator.go [hunk 11]
```diff
@@ -104,26 +130,132 @@ func (state *State) SetDefaultCluster(name string) {
...
+func clusterLookupOnSlice(state *State, clusters []string,
+	resultCh chan<- Result, errorCh chan<- error) {
+	for _, clusterName := range clusters {
 		context := newContext()
-		context.currentClusterName = name
-		for key, _ := range cluster {
+		context.currentClusterName = clusterName
+		for key, _ := range state.clusters[clusterName] {
 			err := clusterLookup(state, &context, key)
+			// we are interested only results for key CLUSTER
```
Comment (xaviershay) [inline evaluator.go]: I think it's because we're doing two things at once: priming the cache for all keys, but also fetching out the CLUSTER results to subsequently pass to populate CQR.

maybe `primeCacheForCluster` or something like that is a more appropriate name, with function documentation explaining that?

Also is there any reason to thread the channels in here? I feel like it needlessly couples - we could just instead return the cluster or an error and have the calling function put them in to the channels?

File: evaluator.go [hunk 12]
```diff
@@ -104,26 +130,132 @@ func (state *State) SetDefaultCluster(name string) {
 // PrimeCache traverses over the entire state to expand all values and store
 // them in the state's cache. Subsequent queries will be able to use the cache
 // immediately, rather than having to build it up incrementally.
-//
+
```
Comment (xaviershay) [inline evaluator.go]: I think this will exclude the top half from the generated documentation, so empty `//` is actually required.

https://godoc.org/github.com/square/grange

I'm pretty sure there's an easy way to generate this locally to verify...

**Comment by xaviershay** [comment]
going to merge on green
